src/api/app/models/cloud/upload_job.rb

```diff
@@ -4,21 +4,55 @@ class UploadJob
     include ActiveModel::Model
     extend Forwardable
 
-    attr_accessor :user_upload_job, :backend_upload_job
-    validate :validate_jobs
+    attr_accessor :user_upload_job, :backend_upload_job, :params
+    validate :validate_jobs, :validate_filename, :validate_architecture, :validate_regions, :validate_virtualization_types, :validate_ami_name
     def_delegator :backend_upload_job, :id
 
     def self.create(user, params)
-      upload_job = new
+      upload_job = new(params: params)
+      return upload_job if upload_job.invalid?
       upload_job.backend_upload_job = Cloud::Backend::UploadJob.create(user, params)
       return upload_job unless upload_job.backend_upload_job.valid?
 
       upload_job.user_upload_job = user.upload_jobs.create(job_id: upload_job.id)
       upload_job
     end
 
+    def self.uploadable_file?(filename)
+      filename.end_with?('.raw.xz', '.vhdfixed.xz')
+    end
+
+    def self.uploadable_architecture?(architecture)
+      architecture == 'x86_64'
+    end
+
     private
 
+    def validate_filename
+      return if self.class.uploadable_file?(params[:filename])
+      errors.add(:base, "'#{params[:filename]}' is not a valid cloud format")
+    end
+
+    def validate_architecture
+      return if self.class.uploadable_architecture?(params[:arch])
+      errors.add(:base, "'#{params[:arch]}' is not a valid cloud format")
+    end
+
+    def validate_regions
+      return if ::Cloud::Ec2::Configuration::REGIONS.map(&:second).include?(params[:region])
+      errors.add(:base, "'#{params[:region]}' is not a valid EC2 region")
+    end
+
+    def validate_virtualization_types
+      return if ::Cloud::Ec2::Configuration::VIRTUALIZATION_TYPES.map(&:second).include?(params[:virtualization_type])
+      errors.add(:base, "'#{params[:virtualization_type]}' is not a valid EC2 virtualization type")
+    end
+
+    def validate_ami_name
+      return if Project.valid_name?(params[:ami_name])
+      errors.add(:base, "'#{params[:ami_name]}' is not a valid EC2 AMI name (only letters, numbers, underscores, hyphens and dots)")
```

Comment:
I don't think you can use params here. And even if, it sounds wrong / misleading

Reply:
Why can't I use it? I assigned it in LOC 12.

Well, the other approach is to assign all values we want to validate to model attributes which means we need to introduce :ami_name, :virtualization_type, :region and :filename.

Reply:
Still it sounds misleading to me. Why not payload or upload_data, upload_configuration_data, ...?